frame/treasury/src/lib.rs

```diff
@@ -299,49 +324,87 @@ impl<T: Trait> Module<T> {
 	}
 
 	/// The needed bond for a proposal whose spend is `value`.
-	fn calculate_bond(value: RingBalance<T>) -> RingBalance<T> {
-		T::ProposalBondMinimum::get().max(T::ProposalBond::get() * value)
+	fn calculate_bonds(ring: RingBalance<T>, kton: KtonBalance<T>) -> (RingBalance<T>, KtonBalance<T>) {
+		let mut ring_bond: RingBalance<T> = RingBalance::<T>::from(0);
+		let mut kton_bond: KtonBalance<T> = KtonBalance::<T>::from(0);
+
+		if ring > ring_bond {
+			ring_bond = T::RingProposalBondMinimum::get().max(T::ProposalBond::get() * ring);
+		}
+
+		if kton > kton_bond {
+			kton_bond = T::KtonProposalBondMinimum::get().max(T::ProposalBond::get() * kton);
+		}
+
+		(ring_bond, kton_bond)
 	}
 
 	// Spend some money!
 	fn spend_funds() {
-		let mut budget_remaining = Self::pot();
-		Self::deposit_event(RawEvent::Spending(budget_remaining));
+		let mut budget_remaining_ring = Self::pot::<T::RingCurrency>();
+		let mut budget_remaining_kton = Self::pot::<T::KtonCurrency>();
+		let mut imbalance_ring = <RingPositiveImbalance<T>>::zero();
+		let mut imbalance_kton = <KtonPositiveImbalance<T>>::zero();
+		let mut miss_any_ring = false;
+		let mut miss_any_kton = false;
+
+		Self::deposit_event(RawEvent::Spending(budget_remaining_ring, budget_remaining_kton));
 
-		let mut missed_any = false;
-		let mut imbalance = <RingPositiveImbalance<T>>::zero();
 		Approvals::mutate(|v| {
 			v.retain(|&index| {
-				// Should always be true, but shouldn't panic if false or we're screwed.
-				if let Some(p) = Self::proposals(index) {
-					if p.value <= budget_remaining {
-						budget_remaining -= p.value;
-						<Proposals<T>>::remove(index);
-
-						// return their deposit.
-						let _ = T::RingCurrency::unreserve(&p.proposer, p.bond);
-
-						// provide the allocation.
-						imbalance.subsume(T::RingCurrency::deposit_creating(&p.beneficiary, p.value));
-
-						Self::deposit_event(RawEvent::Awarded(index, p.value, p.beneficiary));
-						false
-					} else {
-						missed_any = true;
-						true
+				// Should always be some, but shouldn't panic if false or we're screwed.
+				let option_proposal = Self::proposals(index);
+				if option_proposal.is_none() {
+					return false;
+				}
+
+				let p = option_proposal.unwrap();
+				if p.ring_value > budget_remaining_ring {
+					if p.ring_value > budget_remaining_ring {
+						miss_any_ring = true;
 					}
+
+					if p.kton_value > budget_remaining_kton {
+						miss_any_kton = true;
+					}
+
+					return true;
+				}
+
+				if p.ring_value > RingBalance::<T>::from(0) {
+					budget_remaining_ring -= p.ring_value;
```

Comment:
budget_remaining_ring might < p.ring_value, overflow.

The condition checking is wrong too, ```if p.ring_value <= budget_remaining_ring && p.kton_value <= budget_remaining_kton```

Reply:
> budget_remaining_ring might < p.ring_value, overflow.
> 
> The condition checking is wrong too, `if p.ring_value <= budget_remaining_ring && p.kton_value <= budget_remaining_kton`

For here, this is impossible, so this comment can be ignored.

But for the following KTON checking, there are checking issues.